core/client.go

```diff
@@ -0,0 +1,835 @@
+				c.logger.Info("Underlying subscription ended, resubscribing")
+				err := resub()
+				c.handleClientError(err)
+			case <-ctx.Done():
+				c.logger.Info("Context done, ending subscription")
+				safeSub.Unsubscribe()
+				return
+			}
+		}
+	}()
+
+	return safeSub, nil
+}
+
+func (c *SafeEthClient) Close() {
+	c.clientLock.Lock()
+	defer c.clientLock.Unlock()
+
+	close(c.closeC)
+	c.wg.Wait()
+	c.logger.Info("SafeEthClient closed")
+}
+
+func (c *SafeEthClient) isConnectionError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	isConnectionReset := strings.Contains(err.Error(), "connection reset")
+	isConnectionRefused := strings.Contains(err.Error(), "connection refused")
+	isAbnormalClosure := strings.Contains(err.Error(), "abnormal closure")
+
+	return isConnectionReset || isConnectionRefused || isAbnormalClosure
+}
+
+func (c *SafeEthClient) handleClientError(err error) {
+	if err == nil {
+		return
+	}
+
+	if c.isConnectionError(err) {
+		c.logger.Error("Connection error detected, triggering reinit", "err", err)
+		c.triggerReinit()
+	} else {
+		c.logger.Error("Client error detected", "err", err)
+	}
+}
+
+func (c *SafeEthClient) triggerReinit() {
+	c.reinitC <- struct{}{}
+}
+
+func (c *SafeEthClient) SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error) {
+	sub, err := c.Client.SubscribeNewHead(ctx, ch)
+	if err != nil {
+		c.logger.Error("Failed to subscribe to new heads", "err", err)
+		return nil, err
+	}
+	c.logger.Info("Subscribed to new heads")
+
+	safeSub := NewSafeSubscription(sub)
+	intermediateC := make(chan *types.Header, 100)
+
+	resub := func() error {
+		c.clientLock.RLock()
+		defer c.clientLock.RUnlock()
+
+		sub, err = c.Client.SubscribeNewHead(ctx, intermediateC)
+		if err != nil {
+			c.logger.Error("Failed to resubscribe to new heads", "err", err)
+			return err
+		}
+		c.logger.Info("Resubscribed to new heads")
+
+		safeSub.SetUnderlyingSub(sub)
+
+		return nil
+	}
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+
+		reinitC := c.WatchReinit()
+		headerTicker := time.NewTicker(HEADER_TIMEOUT)
+		receivedBlock := false
+
+		for {
+			select {
+			case header := <-intermediateC:
+				receivedBlock = true
+				ch <- header
+			case <-safeSub.Err():
+				c.logger.Info("Safe subscription to new heads ended")
+				return
+			case success := <-reinitC:
+				reinitC = c.WatchReinit()
+				if success {
+					err := resub()
+					c.handleClientError(err)
+				}
+			case <-sub.Err():
+				c.logger.Info("Underlying subscription to new heads ended, resubscribing")
+				err := resub()
+				c.handleClientError(err)
+			case <-headerTicker.C:
+				c.logger.Info("Header ticker fired, ending subscription")
+				if receivedBlock {
+					receivedBlock = false
+				} else {
+					err := resub()
+					c.handleClientError(err)
+				}
+			case <-ctx.Done():
+				c.logger.Info("Context done, ending new heads subscription")
+				safeSub.Unsubscribe()
+				return
+			}
```

Comment:
Tracking `closeC` here as well.

Reply:
Done in [f1f4be6](https://github.com/NethermindEth/near-sffl/pull/154/commits/f1f4be619bfccbcdd5b5dd9e81fc01dfab188a99)